src/libraries/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs

```diff
@@ -143,6 +143,30 @@ public void SetAccessControl_FileInfo_FileSecurity_Success()
             fileInfo.SetAccessControl(fileSecurity);
         }
 
+        [Fact]
+        public void SetAccessControl_FileInfo_FileSecurity_Success_NameLongerThanMaxShortPath()
```

Comment:
@adamsitnik 

I'm sorry I ran it again just now. The test is not throwing exception on my computer neither. But when I copy the test code (with manually adjusted long path) and put it in an empty ConsoleApp, it throws as expected.

Is xUnit environment somewhat special?

Reply:
well, I found VS fail to start debugging for the solution after I updated it hours ago... 😶

Reply:
You are right, it's possible to reproduce it with plain console app, but for some reason not for our test projects.

@ericstj @ViktorHofer @carlossanlop The following code crashes in plain console app:

<details>

```cs
using System.Security.AccessControl;
using System.Security.Principal;

namespace LongPaths
{
    internal class Program
    {
        static void Main(string[] args)
        {
            string tempPath = Path.GetTempPath();

            const int MaxShortPath = 260;
            int fileNameLength = MaxShortPath - tempPath.Length;

            if (fileNameLength < 1)
            {
                fileNameLength = 1;
            }

            string path = Path.Combine(tempPath, new string('1', fileNameLength) + ".txt");
            File.WriteAllBytes(path, new byte[1]);

            try
            {
                FileInfo fileInfo = new(path);
                FileSecurity fileSecurity = fileInfo.GetAccessControl(AccessControlSections.Access);

                FileSystemAccessRule newAccessRule = new(Helpers.s_NetworkServiceNTAccount, FileSystemRights.Write, AccessControlType.Allow);
                fileSecurity.SetAccessRule(newAccessRule);
                fileInfo.SetAccessControl(fileSecurity);
            }
            finally
            {
                File.Delete(path);
            }
        }
    }

    public static class Helpers
    {
        public static IdentityReference s_LocalSystemNTAccount = new SecurityIdentifier(WellKnownSidType.LocalSystemSid, null).Translate(typeof(NTAccount));
        public static IdentityReference s_NetworkServiceNTAccount = new SecurityIdentifier(WellKnownSidType.NetworkServiceSid, null).Translate(typeof(NTAccount));
        public static IdentityReference s_WorldSidNTAccount = new SecurityIdentifier(WellKnownSidType.WorldSid, null);
    }
}
```

</details>

But works fine in our test project. Do you happen to know why?

Reply:
Is it because of:

https://github.com/dotnet/runtime/blob/b82b1d9218b8c68e39a1754c149396bad37a4a49/src/native/corehost/dotnet/dotnet.manifest#L21

?

Reply:
Update: after removing https://github.com/dotnet/runtime/blob/b82b1d9218b8c68e39a1754c149396bad37a4a49/src/native/corehost/dotnet/dotnet.manifest#L21

I was able to repro the issue with our tests

![image](https://github.com/dotnet/runtime/assets/6011991/d485a0b9-127a-4bae-9dbd-d25b216bd19e)

And then verify the fix.

Reply:
cc @agocke @vitek-karas regarding the corehost question above

Reply:
There's a difference in manifests for `dotnet.exe` and the default manifest which one gets when building an app and running it via the executable (we call this the `apphost`).

The `dotnet.exe` added long path support long time ago by default, the `apphost` gets the default manifest produces by the C# compiler (unless it's modified by the developer of the app), which doesn't enable long path support. There's discussion on why it was done and the considerations here: https://github.com/dotnet/core-setup/pull/5335#issuecomment-470706683
One interesting part: 
> We never hit the setting in Core as we use the extended device syntax to get down-level long path support (\\?\). 

I wonder if the new file access APIs don't or can't follow the same pattern - our understanding back then was that if you're using purely .NET APIs for file access, it doesn't matter as those will always be long-path enabled (regardless of the manifest), and that it should only affect code which directly PInvokes into the OS APIs.